### PR TITLE
HHH-18702 Exception using @EmbeddedId with @OneToMany that refers to an alternate key column

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-18702

----------------------
By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt)
and can be relicensed under the terms of the [LGPL v2.1 license](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt) in the future at the maintainers' discretion.
For more information on licensing, please check [here](https://github.com/hibernate/hibernate-orm/blob/main/CONTRIBUTING.md#legal).

----------------------

### CODE REVIEW

**Comment by gavinking** [comment]
Looks reasonable, but is a similar change needed in `EntityType.isEqual()`?

File: hibernate-core/src/main/java/org/hibernate/type/EntityType.java [hunk 1]
```diff
@@ -360,8 +360,15 @@ public int getHashCode(Object x, SessionFactoryImplementor factory) {
 		}
 		else {
 			assert uniqueKeyPropertyName != null;
+			final Object uniqueKey;
 			final Type keyType = persister.getPropertyType( uniqueKeyPropertyName );
-			return keyType.getHashCode( x, factory );
+			if ( keyType.getReturnedClass().isAssignableFrom( x.getClass() ) ) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/type/EntityType.java:365]:
```suggestion
			if ( keyType.getReturnedClass().isInstance( x ) ) {
```

File: hibernate-core/src/main/java/org/hibernate/type/EntityType.java [hunk 2]
```diff
@@ -376,39 +383,62 @@ public boolean isEqual(Object x, Object y, SessionFactoryImplementor factory) {
...
+			// Check for reference equality first as the type-specific checks by IdentifierType are sometimes non-trivial
+			return (xid == yid) || persister.getIdentifierType().isEqual( xid, yid, factory );
 		}
 		else {
-			if ( mappedClass.isAssignableFrom( y.getClass() ) ) {
-				yid = persister.getIdentifier( y );
+			assert uniqueKeyPropertyName != null;
+			final Object xUniqueKey;
+			final Type keyType = persister.getPropertyType( uniqueKeyPropertyName );
+			if ( keyType.getReturnedClass().isAssignableFrom( x.getClass() ) ) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/type/EntityType.java:425]:
```suggestion
			if ( keyType.getReturnedClass().isInstance( x ) ) {
```

File: hibernate-core/src/main/java/org/hibernate/type/EntityType.java [hunk 3]
```diff
@@ -376,39 +383,62 @@ public boolean isEqual(Object x, Object y, SessionFactoryImplementor factory) {
...
-				//JPA 2 case where @IdClass contains the id and not the associated entity
-				yid = y;
+				xUniqueKey = persister.getPropertyValue( x, uniqueKeyPropertyName );
 			}
-		}
 
-		// Check for reference equality first as the type-specific checks by IdentifierType are sometimes non-trivial
-		return ( xid == yid ) || persister.getIdentifierType().isEqual( xid, yid, factory );
+			final Object yUniqueKey;
+			if ( keyType.getReturnedClass().isAssignableFrom( y.getClass() ) ) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/type/EntityType.java:433]:
```suggestion
			if ( keyType.getReturnedClass().isInstance( y ) ) {
```

File: hibernate-core/src/main/java/org/hibernate/type/EntityType.java [hunk 4]
```diff
@@ -376,39 +383,62 @@ public boolean isEqual(Object x, Object y, SessionFactoryImplementor factory) {
...
-		final LazyInitializer lazyInitializerY = extractLazyInitializer( y );
-		if ( lazyInitializerY != null ) {
-			yid = lazyInitializerY.getInternalIdentifier();
+			Object yid;
+			final LazyInitializer lazyInitializerY = extractLazyInitializer( y );
+			if ( lazyInitializerY != null ) {
+				yid = lazyInitializerY.getInternalIdentifier();
+			}
+			else {
+				if ( mappedClass.isAssignableFrom( y.getClass() ) ) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/type/EntityType.java:409]:
```suggestion
				if ( mappedClass.isInstance( y ) ) {
```

File: hibernate-core/src/main/java/org/hibernate/type/EntityType.java [hunk 5]
```diff
@@ -376,39 +383,62 @@ public boolean isEqual(Object x, Object y, SessionFactoryImplementor factory) {
...
+			final Class<?> mappedClass = persister.getMappedClass();
+			Object xid;
+			final LazyInitializer lazyInitializerX = extractLazyInitializer( x );
+			if ( lazyInitializerX != null ) {
+				xid = lazyInitializerX.getInternalIdentifier();
 			}
 			else {
-				//JPA 2 case where @IdClass contains the id and not the associated entity
-				xid = x;
+				if ( mappedClass.isAssignableFrom( x.getClass() ) ) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/type/EntityType.java:394]:
```suggestion
				if ( mappedClass.isInstance( x ) ) {
```

**Review by beikov — APPROVED** [review]
Just a few minor suggestions.

**Pull request merged**